Add plugin P149 - Linear Actuator

rimago opened the pull request:
Hey all,

I have created a simple plugin to control a linear actuator (with integrated stop switches).

It is based on P098 but simplified in the way that it does not require an encoder to set the linear actuator to a specific position.

Instead of using an encoder it uses the time to determine the current position of the linear actuator. You just have to specify the duration of moving from completely in to completely out. It is not fully precise, but simple and sufficient for at least my case of opening a window on certain levels. And if the linear actuator has integrated stop switches, the plugin automatically calibrates the position every time a limit is reached.

If you are interested I am happy to contribute it to the main project. Let me know if changes are necessary.

Best,
 Flo

TD-er commented:
I just browsed over the source code.
But I was wondering, what does this what the other PWM motor controll plugin doesn't? (or can't do)

tonhuisman commented:
A few quick remarks:
- Not sure if differs enough from P098 to be a separate plugin (we're trying to keep code size as small as possible)
- P149 is already reserved, you should have requested a plugin ID [over here](https://github.com/letscontrolit/ESPEasy/issues/3839)

rimago commented:
Thanks for the quick feedback.

This plugin and P098, both, control a motor. However, P098 requires an encoder and two sensors to detect when the limits are reached, whereas this plugin does not require this. Instead it switches on the motor only for a certain time.

I was also thinking about adding the functionality to P098, but did not do that as it would blow up its code. For example, then the encoder, the limits and the timing will be configurable, but only one of it would be used.

What do you think is it better to expand the existing P098 or to create a new small plugin?

Thanks for the hint with the number, I will request a new one, as soon as I know if it will be a separate plugin.

TD-er commented:
The encoder is just optional.
Only difference is you can't move to a specific position if there's no encoder used.

I think just adding a few more commands to the P098 code, like how long it should be turned on, is a lot less code compared to adding a new plugin.

Also if we now add a new plugin, then at some point someone might want a feature added to it, which probably would be nice to have on both as they are so similar.
Thus that will either result in code duplication, or differentiation of features for almost similar plugins.
Both I think are not preferrable.

rimago commented:
Ok, I will give it a try to implement it in P098.

TD-er commented:
Maybe some of the latest features implemented in the "longpulse" command can be useful for you to use here.
You essentially want to start a longpulse.

Another way can be to schedule a `PLUGIN_TASKTIMER_IN` .
This can be called using `setPluginTaskTimer` so you can set a number of int values which will be present in the event when processing `PLUGIN_TASKTIMER_IN` .

rimago commented:
Ok. The idea I have got now is to add a property speed, which is the number of steps per ms. This is used to change the position, when the motor is running. And if an encoder is connected it can be set to 0, which will be the default.

TD-er commented:
Speed, as in "PWM" ? ;)

rimago commented:
:D no. The PWM is the real speed, which slows down the motor. What I mean is a virtual speed that changes the position based on the run duration of the motor. Maybe "virtual speed" is a better name for it ;-)

TD-er commented:
So like the longpulse command with a repeat timer? (fairly recent feature)

rimago commented:
What do you mean with longpulse? I did not find that.

rimago commented:
Ah ok, now, I have found it. Although it might be possible to control the motor with that, it is not really useful.

I want to be able to control the motor with the movetopos command. This requires that the plugin is aware of an estimate of the current position of the motor and moves it only the required difference. Furthermore to move the motor I have to control two GPIO pins.

TD-er commented:
> Ah ok, now, I have found it. Although it might be possible to control the motor with that, it is not really useful.

I know and it was by no means my intention to let you use the LongPulse command from rules as that may cause too much uncontrollable delays in processing.
But you can have a look at the special (C++) function calls used when handling that command, so you only need to make a single function call to implement it.
And processing the command from P098 is probably very much like processing that longpulse command, so you can get even more "inspiration" from that code :)

rimago committed:
Enhance P098 to work without encoder and limit sensors

* add virtual speed to estimate position without encoder
* add position 0 supplement to calibrate position, when moving to position 0
* make publishable to controllers
  - publish values in PLUGIN_READ
  - set SendDataOption = true
  - set TimerOption = true

Signed-off-by: Florian Binder <fb@java4.info>

rimago commented:
I will create a new pull request for the enhancement of P098.
